File: mvp/runtime-hosts/goose-host-v2/src/grant_channel.rs
use std::collections::BTreeSet;
use std::fs::File;
use std::io::{Read, Write};
use std::os::fd::{FromRawFd, RawFd};
use std::thread::{self, JoinHandle};
use std::time::{SystemTime, UNIX_EPOCH};

use serde::Deserialize;
use serde_json::{Value, json};
use sha2::{Digest, Sha256};

const GRANT_MAGIC: &[u8; 8] = b"JAGTGRN1";
const ACK_MAGIC: &[u8; 8] = b"JAGTACK1";
const WIRE_VERSION: u8 = 1;
const MAX_HEADER_BYTES: usize = 65_536;
const MAX_SECRET_BYTES: usize = 65_536;
const MAX_ACK_BYTES: usize = 8_192;

pub const PROVIDER_GRANT_FD_ENV: &str = "WORKBENCH_PROVIDER_GRANT_FD";

pub fn validate_preopened_fd(fd: RawFd) -> Result<(), String> {
    if fd <= 2 {
        return Err("provider Grant Channel must use a private pre-opened descriptor".into());
    }
    Ok(())
}

fn restore_close_on_exec(fd: RawFd) -> Result<(), String> {
    validate_preopened_fd(fd)?;
    // SAFETY: fcntl only observes and updates flags on the validated descriptor.
    let flags = unsafe { libc::fcntl(fd, libc::F_GETFD) };
    if flags < 0 {
        return Err("provider Grant Channel descriptor is unavailable".into());
    }
    // SAFETY: the descriptor remains owned by this process until read_once.
    if unsafe { libc::fcntl(fd, libc::F_SETFD, flags | libc::FD_CLOEXEC) } < 0 {
        return Err("provider Grant Channel containment failed".into());
    }
    Ok(())
}

pub struct GrantReceiver {
    task: Option<JoinHandle<Result<GrantMaterial, String>>>,
}

impl GrantReceiver {
    pub fn start(fd: RawFd) -> Result<Self, String> {
        restore_close_on_exec(fd)?;
        Ok(Self {
            task: Some(thread::spawn(move || read_once(fd))),
        })
    }

    pub fn receive(&mut self) -> Result<GrantMaterial, String> {
        let task = self
            .task
            .take()
            .ok_or("provider Grant Channel was already consumed")?;
        task.join()
            .map_err(|_| "provider Grant Channel receiver failed".to_owned())?
    }
}

pub struct GrantMaterial {
    binding: ProviderGrantBinding,
    bytes: Vec<u8>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FixtureDisposition {
    Complete,
    Fail,
    Hold,
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
struct ProviderGrantTarget {
    runtime_id: String,
    build_id: String,
    lease_id: String,
    instance_id_digest: String,
    instance_nonce_digest: String,
    host_generation: String,
    lease_generation_seq: u64,
    expires_at: f64,
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
struct ProviderGrantBinding {
    grant_id: String,
    target: ProviderGrantTarget,
    session_id: String,
    command_id: String,
    run_id: String,
    term_id: String,
    step_id: String,
    provider_id: String,
    provider_profile_digest: String,
    model: String,
    scopes: Vec<String>,
    issued_at: f64,
    expires_at: f64,
    grant_nonce_digest: String,
}

impl GrantMaterial {
    pub fn is_empty(&self) -> bool {
        self.bytes.is_empty()
    }

    #[allow(clippy::too_many_arguments)]
    pub fn fixture_disposition(
        &self,
        command_id: &str,
        run_id: &str,
        term_id: &str,
        step_id: &str,
        provider_ref: &str,
        _model_alias: &str,
    ) -> Result<FixtureDisposition, String> {
        let expected_provider_ref = format!("provider-profile:{}", self.binding.provider_id);
        if self.binding.command_id != command_id
            || self.binding.run_id != run_id
            || self.binding.term_id != term_id
            || self.binding.step_id != step_id
            || expected_provider_ref != provider_ref
            || self.binding.model.is_empty()
        {
            return Err("provider Grant binding identity mismatch".into());
        }
        match self.binding.provider_id.as_str() {
            "fixture" | "fixture-completed" => Ok(FixtureDisposition::Complete),
            "fixture-failed" => Ok(FixtureDisposition::Fail),
            "fixture-held" => Ok(FixtureDisposition::Hold),
            _ => Err("fixed Goose provider is unsupported".into()),
        }
    }
}

impl Drop for GrantMaterial {
    fn drop(&mut self) {
        self.bytes.fill(0);
    }
}

fn read_once(fd: RawFd) -> Result<GrantMaterial, String> {
    validate_preopened_fd(fd)?;
    // SAFETY: the Supervisor transfers ownership of this pre-opened descriptor
    // to the sidecar. It is consumed once and closed when `file` is dropped.
    let mut file = unsafe { File::from_raw_fd(fd) };
    let mut prefix = [0_u8; 17];
    file.read_exact(&mut prefix)
        .map_err(|_| "provider Grant Channel framing is incomplete")?;
    if &prefix[..8] != GRANT_MAGIC || prefix[8] != WIRE_VERSION {
        return Err("provider Grant Channel framing is invalid".into());
    }
    let header_size = u32::from_be_bytes(prefix[9..13].try_into().expect("header size")) as usize;
    let secret_size = u32::from_be_bytes(prefix[13..17].try_into().expect("secret size")) as usize;
    if header_size == 0
        || header_size > MAX_HEADER_BYTES
        || secret_size == 0
        || secret_size > MAX_SECRET_BYTES
    {
        return Err("provider Grant Channel framing is invalid".into());
    }
    let mut header_bytes = vec![0_u8; header_size];
    file.read_exact(&mut header_bytes)
        .map_err(|_| "provider Grant Channel header is incomplete")?;
    let (binding, grant_digest) = parse_header(&header_bytes)?;
    header_bytes.fill(0);
    let mut secret = vec![0_u8; secret_size];
    if file.read_exact(&mut secret).is_err() {
        secret.fill(0);
        return Err("provider Grant Channel secret is incomplete".into());
    }
    let acknowledgement = json!({
        "schema":"workbench.runtime.provider_grant_ack.v1",
        "grant_id":binding.grant_id,
        "grant_digest":grant_digest,
        "target_instance_digest":binding.target.instance_id_digest,
    });
    let ack = serde_json::to_vec(&acknowledgement)
        .map_err(|_| "provider Grant Channel acknowledgement failed")?;
    if ack.is_empty() || ack.len() > MAX_ACK_BYTES {
        secret.fill(0);
        return Err("provider Grant Channel acknowledgement failed".into());
    }
    let mut ack_prefix = Vec::with_capacity(13);
    ack_prefix.extend_from_slice(ACK_MAGIC);
    ack_prefix.push(WIRE_VERSION);
    ack_prefix.extend_from_slice(&(ack.len() as u32).to_be_bytes());
    if file
        .write_all(&ack_prefix)
        .and_then(|_| file.write_all(&ack))
        .is_err()
    {
        secret.fill(0);
        return Err("provider Grant Channel acknowledgement failed".into());
    }
    file.flush()
        .map_err(|_| "provider Grant Channel acknowledgement failed")?;
    Ok(GrantMaterial {
        binding,
        bytes: secret,
    })
}

fn parse_header(bytes: &[u8]) -> Result<(ProviderGrantBinding, String), String> {
    let document: Value =
        serde_json::from_slice(bytes).map_err(|_| "provider Grant Channel header is invalid")?;
    let object = document
        .as_object()
        .ok_or("provider Grant Channel header is invalid")?;
    if object.len() != 3
        || object.get("schema").and_then(Value::as_str)
            != Some("workbench.runtime.provider_grant_private.v1")
    {
        return Err("provider Grant Channel header is invalid".into());
    }
    let grant_digest = object
        .get("grant_digest")
        .and_then(Value::as_str)
        .filter(|value| is_digest(value))
        .ok_or("provider Grant Channel header is invalid")?
        .to_owned();
    let binding_value = object
        .get("binding")
        .cloned()
        .ok_or("provider Grant Channel header is invalid")?;
    if canonical_value_digest(binding_value.clone())? != grant_digest {
        return Err("provider Grant Channel header is invalid".into());
    }
    let binding: ProviderGrantBinding = serde_json::from_value(binding_value)
        .map_err(|_| "provider Grant Channel binding is invalid")?;
    validate_binding(&binding)?;
    Ok((binding, grant_digest))
}

fn validate_binding(binding: &ProviderGrantBinding) -> Result<(), String> {
    let target = &binding.target;
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_err(|_| "provider Grant Channel clock is invalid")?
        .as_secs_f64();
    let scopes: BTreeSet<&str> = binding.scopes.iter().map(String::as_str).collect();
    if target.runtime_id != "goose"
        || target.build_id != "goose-host-v2:fixture-wrapper-r2"
        || target.lease_generation_seq == 0
        || !is_digest(&target.instance_id_digest)
        || !is_digest(&target.instance_nonce_digest)
        || !is_digest(&binding.provider_profile_digest)
        || !is_digest(&binding.grant_nonce_digest)
        || binding.scopes.is_empty()
        || scopes.len() != binding.scopes.len()
        || binding.expires_at <= now
        || binding.expires_at > target.expires_at
        || binding.issued_at <= 0.0
        || binding.expires_at <= binding.issued_at
        || [
            &binding.grant_id,
            &binding.session_id,
            &binding.command_id,
            &binding.run_id,
            &binding.term_id,
            &binding.step_id,
            &binding.provider_id,
            &binding.model,
            &target.lease_id,
            &target.host_generation,
        ]
        .iter()
        .any(|value| value.is_empty())
        || binding.provider_id.contains([':', '/'])
    {
        return Err("provider Grant Channel binding is invalid or expired".into());
    }
    Ok(())
}

fn canonical_value_digest(value: Value) -> Result<String, String> {
    let encoded = serde_json::to_vec(&canonical(value))
        .map_err(|_| "provider Grant Channel binding is invalid")?;
    Ok(format!("{:x}", Sha256::digest(encoded)))
}

fn canonical(value: Value) -> Value {
    match value {
        Value::Array(values) => Value::Array(values.into_iter().map(canonical).collect()),
        Value::Object(values) => Value::Object(
            values
                .into_iter()
                .map(|(key, value)| (key, canonical(value)))
                .collect(),
        ),
        scalar => scalar,
    }
}

fn is_digest(value: &str) -> bool {
    value.len() == 64
        && value
            .bytes()
            .all(|byte| byte.is_ascii_hexdigit() && !byte.is_ascii_uppercase())
}

#[cfg(test)]
mod tests {
    use super::validate_preopened_fd;

    #[test]
    fn grant_channel_must_be_private_and_preopened() {
        assert!(validate_preopened_fd(3).is_ok());
        assert!(validate_preopened_fd(0).is_err());
        assert!(validate_preopened_fd(1).is_err());
        assert!(validate_preopened_fd(2).is_err());
    }
}

File: mvp/sidecars/deepseek-harness/src/grant-channel.ts
import { createHash } from "node:crypto";
import { Socket } from "node:net";


const DIGEST = /^[0-9a-f]{64}$/;
const BINDING_IDENTITY_FIELDS = Object.freeze([
  "command_id", "run_id", "term_id", "step_id", "provider_ref",
]);
const GRANT_HEADER_FIELDS = Object.freeze(["schema", "binding", "grant_digest"]);
const GRANT_BINDING_FIELDS = Object.freeze([
  "grant_id", "target", "session_id", "command_id", "run_id", "term_id", "step_id",
  "provider_id", "provider_profile_digest", "model", "scopes", "issued_at", "expires_at",
  "grant_nonce_digest",
]);
const GRANT_TARGET_FIELDS = Object.freeze([
  "runtime_id", "build_id", "lease_id", "instance_id_digest", "instance_nonce_digest",
  "host_generation", "lease_generation_seq", "expires_at",
]);
const GRANT_MAGIC = Buffer.from("JAGTGRN1", "ascii");
const ACK_MAGIC = Buffer.from("JAGTACK1", "ascii");
const WIRE_VERSION = 1;
const GRANT_PREFIX_BYTES = 17;
const MAX_HEADER_BYTES = 65_536;
const MAX_SECRET_BYTES = 65_536;
const MAX_ACK_BYTES = 8_192;


function exactKeys(value, expected) {
  return value !== null && typeof value === "object" && !Array.isArray(value)
    && Object.keys(value).sort().join(",") === [...expected].sort().join(",");
}


function canonical(value) {
  if (Array.isArray(value)) return value.map(canonical);
  if (value !== null && typeof value === "object") {
    return Object.fromEntries(Object.keys(value).sort().map(key => [key, canonical(value[key])]));
  }
  return value;
}


function canonicalDigest(value) {
  return createHash("sha256").update(JSON.stringify(canonical(value))).digest("hex");
}


export class EphemeralGrantChannel {
  constructor(clock = () => Date.now() / 1000, targetInstanceDigest = null) {
    this.clock = clock;
    this.targetInstanceDigest = targetInstanceDigest;
    this.grants = new Map();
  }

  accept(binding, transientSecret) {
    if (!binding || typeof binding.grant_id !== "string" || binding.grant_id.length === 0
        || !DIGEST.test(binding.grant_digest)
        || !DIGEST.test(binding.target_instance_digest)
        || (this.targetInstanceDigest !== null
          && binding.target_instance_digest !== this.targetInstanceDigest)
        || typeof binding.expires_at !== "number"
        || !Number.isFinite(binding.expires_at)
        || binding.expires_at <= this.clock()
        || BINDING_IDENTITY_FIELDS.some(
          field => typeof binding[field] !== "string" || binding[field].length === 0,
        )) {
      throw new Error("provider grant binding is invalid or expired");
    }
    if (!Buffer.isBuffer(transientSecret) || transientSecret.length === 0) {
      throw new Error("provider grant secret is invalid");
    }
    if (this.grants.has(binding.grant_id)) {
      throw new Error("provider grant replay is forbidden");
    }
    const retained = Buffer.from(transientSecret);
    transientSecret.fill(0);
    this.grants.set(binding.grant_id, { binding: Object.freeze({ ...binding }), secret: retained });
    return Object.freeze({
      grant_id: binding.grant_id,
      grant_digest: binding.grant_digest,
      target_instance_digest: binding.target_instance_digest,
      acknowledged_at: this.clock(),
    });
  }

  consumeForEnvelope(envelope) {
    if (!envelope || BINDING_IDENTITY_FIELDS.some(
      field => typeof envelope[field] !== "string" || envelope[field].length === 0,
    )) {
      throw new Error("provider grant envelope identity is invalid");
    }
    for (const [grantId, entry] of this.grants) {
      if (entry.binding.expires_at <= this.clock()) {
        entry.secret.fill(0);
        this.grants.delete(grantId);
      }
    }
    const matching = [...this.grants.entries()].filter(([, entry]) => (
      BINDING_IDENTITY_FIELDS.every(field => entry.binding[field] === envelope[field])
    ));
    if (matching.length !== 1) {
      throw new Error("provider grant is unavailable");
    }
    const [grantId, entry] = matching[0];
    this.grants.delete(grantId);
    return Object.freeze({
      secret: entry.secret,
      acknowledgement: Object.freeze({
        grant_id: entry.binding.grant_id,
        grant_digest: entry.binding.grant_digest,
        target_instance_digest: entry.binding.target_instance_digest,
        command_id: entry.binding.command_id,
        run_id: entry.binding.run_id,
        term_id: entry.binding.term_id,
        step_id: entry.binding.step_id,
        resolved_model: entry.binding.model,
        acknowledged_at: this.clock(),
      }),
    });
  }

  clear() {
    for (const { secret } of this.grants.values()) secret.fill(0);
    this.grants.clear();
  }

  static digest(binding) {
    return canonicalDigest(binding);
  }
}


function validateFormalGrant(document, targetIdentity, clock) {
  if (!exactKeys(document, GRANT_HEADER_FIELDS)
      || document.schema !== "workbench.runtime.provider_grant_private.v1"
      || !DIGEST.test(document.grant_digest)
      || !exactKeys(document.binding, GRANT_BINDING_FIELDS)
      || !exactKeys(document.binding.target, GRANT_TARGET_FIELDS)
      || canonicalDigest(document.binding) !== document.grant_digest) {
    throw new Error("private provider grant header is invalid");
  }
  const binding = document.binding;
  const target = binding.target;
  if (target.runtime_id !== targetIdentity.runtimeId
      || target.build_id !== targetIdentity.buildId
      || !DIGEST.test(target.instance_id_digest)
      || !DIGEST.test(target.instance_nonce_digest)
      || !Number.isSafeInteger(target.lease_generation_seq)
      || target.lease_generation_seq < 1
      || !Number.isFinite(target.expires_at)
      || typeof binding.grant_id !== "string" || binding.grant_id.length === 0
      || typeof binding.provider_id !== "string" || binding.provider_id.length === 0
      || binding.provider_id.includes(":") || binding.provider_id.includes("/")
      || !DIGEST.test(binding.provider_profile_digest)
      || !DIGEST.test(binding.grant_nonce_digest)
      || typeof binding.model !== "string" || binding.model.length === 0
      || !Array.isArray(binding.scopes) || binding.scopes.length === 0
      || binding.scopes.some(scope => typeof scope !== "string" || scope.length === 0)
      || new Set(binding.scopes).size !== binding.scopes.length
      || !Number.isFinite(binding.issued_at)
      || !Number.isFinite(binding.expires_at)
      || binding.expires_at <= clock()
      || binding.expires_at > target.expires_at
      || ["session_id", "command_id", "run_id", "term_id", "step_id"].some(
        field => typeof binding[field] !== "string" || binding[field].length === 0,
      )) {
    throw new Error("private provider grant binding is invalid or expired");
  }
  return binding;
}


function ackFrame(binding, grantDigest) {
  const payload = Buffer.from(JSON.stringify({
    schema: "workbench.runtime.provider_grant_ack.v1",
    grant_id: binding.grant_id,
    grant_digest: grantDigest,
    target_instance_digest: binding.target.instance_id_digest,
  }), "utf8");
  if (payload.length > MAX_ACK_BYTES) {
    throw new Error("private provider grant acknowledgement is too large");
  }
  const prefix = Buffer.alloc(13);
  ACK_MAGIC.copy(prefix, 0);
  prefix.writeUInt8(WIRE_VERSION, 8);
  prefix.writeUInt32BE(payload.length, 9);
  return Buffer.concat([prefix, payload]);
}


export function startPreopenedGrantReceiver(
  fd,
  targetIdentity,
  clock = () => Date.now() / 1000,
) {
  if (!Number.isSafeInteger(fd) || fd <= 2
      || targetIdentity === null || typeof targetIdentity !== "object"
      || typeof targetIdentity.runtimeId !== "string" || targetIdentity.runtimeId.length === 0
      || typeof targetIdentity.buildId !== "string" || targetIdentity.buildId.length === 0) {
    throw new Error("private provider grant descriptor is invalid");
  }
  const channel = new EphemeralGrantChannel(clock);
  let endpoint;
  try {
    endpoint = new Socket({ fd, readable: true, writable: true });
  } catch {
    throw new Error("private provider grant descriptor is unavailable");
  }
  let buffered = Buffer.alloc(0);
  let expectedBytes = null;
  const ready = new Promise((resolve, reject) => {
    const fail = message => {
      channel.clear();
      buffered.fill(0);
      endpoint.destroy();
      reject(new Error(message));
    };
    endpoint.on("error", () => fail("private provider grant descriptor is unavailable"));
    endpoint.on("end", () => {
      if (expectedBytes === null || buffered.length !== expectedBytes) {
        fail("private provider grant payload is incomplete");
      }
    });
    endpoint.on("data", chunk => {
      buffered = Buffer.concat([buffered, chunk]);
      if (buffered.length >= GRANT_PREFIX_BYTES && expectedBytes === null) {
        if (!buffered.subarray(0, 8).equals(GRANT_MAGIC)
            || buffered.readUInt8(8) !== WIRE_VERSION) {
          fail("private provider grant framing is invalid");
          return;
        }
        const headerBytes = buffered.readUInt32BE(9);
        const secretBytes = buffered.readUInt32BE(13);
        if (headerBytes === 0 || headerBytes > MAX_HEADER_BYTES
            || secretBytes === 0 || secretBytes > MAX_SECRET_BYTES) {
          fail("private provider grant framing is invalid");
          return;
        }
        expectedBytes = GRANT_PREFIX_BYTES + headerBytes + secretBytes;
      }
      if (expectedBytes === null || buffered.length < expectedBytes) return;
      if (buffered.length !== expectedBytes) {
        fail("private provider grant contains trailing bytes");
        return;
      }
      const headerBytes = buffered.readUInt32BE(9);
      let document;
      try {
        document = JSON.parse(
          buffered.subarray(GRANT_PREFIX_BYTES, GRANT_PREFIX_BYTES + headerBytes).toString("utf8"),
        );
      } catch {
        fail("private provider grant header is invalid");
        return;
      }
      let binding;
      try {
        binding = validateFormalGrant(document, targetIdentity, clock);
      } catch (error) {
        fail(error instanceof Error ? error.message : "private provider grant is invalid");
        return;
      }
      const secret = buffered.subarray(GRANT_PREFIX_BYTES + headerBytes);
      let acknowledgement;
      try {
        acknowledgement = channel.accept({
          grant_id: binding.grant_id,
          grant_digest: document.grant_digest,
          target_instance_digest: binding.target.instance_id_digest,
          command_id: binding.command_id,
          run_id: binding.run_id,
          term_id: binding.term_id,
          step_id: binding.step_id,
          provider_ref: `provider-profile:${binding.provider_id}`,
          model: binding.model,
          expires_at: binding.expires_at,
        }, secret);
        const frame = ackFrame(binding, acknowledgement.grant_digest);
        buffered.fill(0);
        endpoint.end(frame, () => {
          endpoint.destroy();
          resolve(channel);
        });
      } catch {
        secret.fill(0);
        fail("private provider grant payload is invalid");
      }
    });
  });
  // Mark the background failure as observed; query.start still awaits the same
  // promise and reports the closed failure through Host v2.
  ready.catch(() => {});
  return Object.freeze({ channel, ready });
}


export function readPreopenedGrantChannel(fd, targetIdentity, clock) {
  const receiver = startPreopenedGrantReceiver(fd, targetIdentity, clock);
  return receiver;
}
